### DISCUSSION TITLE
Accounts improvements suggestions 👥

### DISCUSSION THREAD

**Discussion by StorytellerCZ** [meteor/meteor#11653] [Feature Requests]
Hello everyone! Accounts are a central part of Meteor and we are striving to bring in new functionality and improvements to make it easier to get started with you app and accounts.

We have a new login method `passwordless` coming up in Meteor 2.5:
Discussion: https://github.com/meteor/meteor/discussions/11515
PR: https://github.com/meteor/meteor/pull/11602

Following that we have two improvements in configuring accounts:
Allow setting `Accounts.config` from Meteor settings: https://github.com/meteor/meteor/pull/11649
Set login services from Meteor settings: https://github.com/meteor/meteor/pull/11650

We are also planing to add 2FA for accounts: https://github.com/meteor/meteor/discussions/11514

And there is also an idea to bring to accounts linking to core: https://github.com/meteor/meteor/discussions/11516
Thought that is still up for a debate.

Bearing those improvements in mind I would like ask you to suggest other improvements to the accounts system. Right now we are mainly looking for things that would make your life easier when setting accounts up so that you don't have to spend more time/code on it and move on to other parts of your app. Use the examples of what we already have in works to inspire you or suggest something awesome.

**Comment by rj-david** [comment]
Support for webauthn: https://webauthn.guide/

On a related note, using `passwordless` above seems like the wrong term to use. Technically, all token-based authentications (webauthn, oauth, one-link, etc.) are `passwordless`.

**Reply by filipenevola** [reply]
> using `passwordless` above seems like the wrong term to use

@rj-david what do you think would be a better package name?

**Reply by distalx** [reply]
Maybe `single-factor` authentication!!

**Reply by guncebektas** [reply]
What about authWithEmailToken ?

**Reply by StorytellerCZ** [reply]
Probably best to move the discussion about passwordless name to that discussion: https://github.com/meteor/meteor/discussions/11515

As for the WebAuth, I'm wondering if that could not be plugged into the existing structures. I'm interested in things like these, but I'm not sure if that is a candidate for inclusion in core. More feedback on this is appreciated.

**Reply by jamauro** [reply]
+1 for WebAuthn support. Going forward I anticipate apps going in this direction rather than using the existing methods. Maybe it can be plugged into the existing structures (I’m not sure) but certainly doesn’t work out of the box like other methods.

**Reply by rj-david** [reply]
webauthn has become more powerful now with passkeys implementation from both Apple and Google

**Comment by Floriferous** [comment]
This is really off the top of my head, but meteor accounts emails have always been a bit weird to manage. The API is weird, and using our own email provider (like mandrill, or others) feels like jumping through hoops. 

Maybe an improved, more granular API, with opt-in (or opt-out) of the default email setup would be helpful.

A couple examples
* If you need additional data from DB to format the Subject and text of an email, you need to query your DB twice in the custom `subject()` and `text()` methods.
* If you want to add a file to an email, you need to hack the `Accounts.generateOptionsForEmail` method.

It's all doable, and it works, but the patterns feel very old and clunky.

**Reply by StorytellerCZ** [reply]
:thinking:

For the first example I'm thinking of adding one method which would get the usual variables and would expect a return of an object with subject, text and html. Possibly even other things like from field to have a full email object. That would also solve the second case with adding a file to email if we go with the route of having to return NodeMailer object.

I'll add making a review of accounts related email functionality to my todo list.

This is a good feedback! Keep it coming!

**Comment by jkasevits** [comment]
Accounts are a major benefit of Meteor, so I'm glad to see it is receiving some attention.

I myself would like to see `Accounts.registerLoginHandler` and `Accounts.callLoginMethod` documented in the Meteor docs. These two tools make the accounts system very flexible and easy to work with. But since they are not documented, many won't find these. And from those that do, some will choose not to make use of these out of fear that it's not documented API and may change in the future without notice. A small reference in the docs would take minimal effort and provide substantial benefits.

These are especially useful for implementing custom multi-factor authentication flows. For example like described in [this Meteor blog post](https://blog.meteor.com/tutorial-two-factor-authentication-with-meteor-and-totp-21d4a2f9ee51). But not only MFA. For a new app I'm currently building I need a signup and login flow that is fully based on users' phone numbers (no usernames, emails or password). Users input their phone number, receive a SMS with a one time code and use that for signup/login. For some public facing apps this is a perfect balance between convenience and security.

EDIT: I might have been a bit quick about writing my thoughts. It seems the new `passwordless` login method coming in 2.5 is the same thing that I need currently, just works via email instead of SMS. However, I still think it would be beneficial to document `Accounts.registerLoginHandler` and `Accounts.callLoginMethod`. Even with new ready-to-use login methods added, flexibility is still beneficial.

**Comment by pmogollons** [comment]
Something that we did had some issues was creating a user in a transaction. With the current implementation of Accounts.createUser is not possible to create users using a session, our use case is for when creating users in batch from a csv or an api call.

**Comment by rafahoro** [comment]
Hi,
  i would like to request for support/clarify sign-in with apple (for Cordova apps)
Not sure if is specific for de Accounts package or may be for the https://github.com/Meteor-Community-Packages/meteor-link-accounts, but will be great to have some 'official' package/way of adding Apple-Sign-In to meteor/Cordova apps and document it in the guides.
Googling it, I've found some packages, but not sure which one to use (and https://guide.meteor.com/accounts.html doesn't mention neither).
I guess these are the ones, but not sure:
 - https://github.com/Meteor-Community-Packages/meteor-link-accounts/pull/89
 - https://github.com/quavedev/apple-oauth

**Comment by xet7** [comment]
Hi,
I'm maintainer of WeKan ® is Open Source Kanban https://wekan.github.io and https://github.com/wekan/wekan . Wekan is made with Meteor, using Node.js and MongoDB. [Related Meteor upgrade issue here](https://github.com/wekan/wekan/issues/3881).

Listening about Meteor Impact talk 2021-10-21 [Meteor Accounts 2.0 and Beyond](https://impact.meteor.com/meetings/virtual/Pyem5vNjDGWRP9jHd), what are the current newest mainained packages for these following? Related code changes have links at [Wekan ChangeLog](https://github.com/wekan/wekan/blob/master/CHANGELOG.md).

1) OAuth2/OIDC/Auth0, that is all based on forked and modified OIDC/OAuth2 package. For these, I don't know do these ways to login use all OIDC specs correctly. Wekan has code of it at https://github.com/wekan/wekan/tree/master/packages/wekan-accounts-oidc and https://github.com/wekan/wekan/tree/master/packages/wekan-accounts-oidc and https://github.com/wekan/wekan/blob/master/client/components/main/layouts.js#L235-L313 and https://github.com/wekan/wekan/blob/master/server/authentication.js and Admin Panel to support [OAuth2 for RocketChat, GitLab, LemonLDAP::NG](https://github.com/wekan/wekan/wiki/OAuth2), [Auth0](https://github.com/wekan/wekan/wiki/OAuth2#auth0), [Azure](https://github.com/wekan/wekan/wiki/Azure), [ADFS 4.0](https://github.com/wekan/wekan/wiki/ADFS), [Oracle OIM](https://github.com/wekan/wekan/wiki/Oracle-OIM), [Keycloak](https://github.com/wekan/wekan/wiki/Keycloak) (Keycloak maybe works or not), [Nextcloud](https://github.com/wekan/wekan/wiki/Nextcloud), [Google login](https://github.com/wekan/wekan/wiki/Google-login).

2) LDAP. Wekan has code of it at https://github.com/wekan/wekan/tree/master/packages/wekan-ldap . LDAP based config is at wiki at [LDAP](https://github.com/wekan/wekan/wiki/LDAP) and [LDAP AD Simple Auth](https://github.com/wekan/wekan/wiki/LDAP-AD-Simple-Auth). LDAP is used in Wekan source/Docker/Snap versions.

3) CAS. Wekan has code of it here https://github.com/wekan/wekan/tree/master/packages/meteor-accounts-cas and wiki page at https://github.com/wekan/wekan/wiki/CAS , but it's not tested does it currently work or is there newer CAS somewhere. Also RocketChat seems to have some CAS code.

4) SAML. For Standalone Wekan, there used to be [some SAML package](https://github.com/wekan/wekan/wiki/SAML) but I don't find it currently. RocketChat has some more up-to-date SAML code in Community version. Where is newest SAML?

5) For [Sandstorm](https://sandstorm.io) , Wekan uses `kenton:accounts-sandstorm` that has it's own SSO of Google/GitHub/LDAP/SAML/Passwordless email login for all Sandstorm Grain Apps.

6) Header login, is there something implemented yet for Meteor with any package? https://github.com/wekan/wekan/issues/2019 . At Sandstorm platform Wekan can autologin using Sandstorm header https://github.com/wekan/wekan/blob/master/sandstorm.js#L370 , but there is not that feature for Standalone Wekan (Source/Docker/Snap) yet.

Current Wekan does have [Teams/Organizations](https://github.com/wekan/wekan/issues/802) at Admin Panel, Boards and All Boards pages, but it's kind of added at top of Meteor Accounts package so it works both in Standalone Wekan (Source/Docker/Snap), it is
not integrated into Sandstorm SSO accounts in any way.

Wekan also has some permissions for Boards users like BoardAdmin/NoComments/CommentOnly/Worker, but it's all custom hardcoded, and not based on `alanning:roles` currently.

**Reply by StorytellerCZ** [reply]
Hi @xet7! Thank you for your exhausting post
I will look through the packages to see if I can see some that will work. Some of these I think will be interesting to consider in integrating into code. Specifially LDAP and SAML have caught my attention that we should look into introducing into core due to its use in corporate world.
More to come on this from me after I recover from Impact.

**Comment by cormip** [comment]
I've just implemented the passwordless login and discovered that the 6-character code emailed to the user appears to use uppercase letters and numbers. It also uses characters that are easily confused like uppercase "O" and zero "0", also number "1" and uppercase "I". I would suggest using only digits (much easier for short-term recall) like most of the similar authentication schemes use or NOT using the ambiguous characters at all when generating the code. Ideally, allowing the code complexity to be configurable (to appease some apps' security requirements) would be ideal.

**Comment by michaelcbrook** [comment]
These two items aren't necessarily ease-of-use related, but security-related items having to do with Accounts. Just wanted to put these here so they remain on the radar, as I feel they will become significant.

https://github.com/meteor/meteor/discussions/11759: Add PKCE support for OAuth (required as of OAuth 2.1)
https://github.com/meteor/meteor/discussions/11812: Password shucking vulnerability

**Reply by StorytellerCZ** [reply]
Argon2 migration update here: https://github.com/meteor/meteor/discussions/13529#discussioncomment-11869775

**Comment by harveysanders** [comment]
What do we think about adding a [post account creation hook?](https://github.com/meteor/meteor/discussions/13637)

**Reply by StorytellerCZ** [reply]
That is what [onCreateUser](https://docs.meteor.com/api/accounts.html#AccountsServer-onCreateUser) pretty much is.

**Reply by harveysanders** [reply]
We're using `onCreateUser`, but it's not ideal because it's called _before_ insertion and assumes that the insertion will always succeed. Granted, the insertion will succeed most of the time, but if you're creating/updating documents in other collections based on the success of a new user and the insert fails, you're kind of out of luck.
We could use a Mongo transaction, but AFAICT, there is no clear hook for when the account creation fails

**Reply by StorytellerCZ** [reply]
Continued here: https://github.com/meteor/meteor/discussions/13637
